usr/man/man8/radisk.8
.\" to produce output: 
.\" ditroff -Tln01 -t -man filename | lpr -n -Pln &
.TH radisk 8 dc
.UC 4
.SH NAME
radisk \- DSA disk maintenance
.SH SYNTAX
.B /etc/radisk
.sp
.B radisk
.B \-c 
.I LBN
.I length
.I special
.RE
.br
.B radisk
.B \-r
.I LBN
.I special
.br
.B radisk
.B \-s
.I LBN
.I length
.I special
.SH DESCRIPTION
The DSA disk maintenance program
.PN radisk
provides the essential functions to manage DSA (DIGITAL Storage
Architecture) disk devices. The
.PN radisk
command must be used on unmounted disk partitions to insure correct
results.
.PP
The
.I LBN 
is a decimal number that represents the logical block number as
reported in the errorlog file.  The LBN is the actual disk block number
starting from the beginning of the disk. The LBN specified with the
.B \-s 
option can be 
.B 0 
to indicate the first block in the specified partition.
.PP
The
.I length 
is a decimal number that indicates how many (512 byte) blocks to
process.  The length specified may be 
.B \-1 
to indicate the last block of the specified partition.
.PP
For the 
.B \-c 
and 
.B \-r 
options, the
.I special
file specified indicates an unmounted 
.B c
partition of a character device special file.  The 
.B \-s 
option will accept any valid partition on the disk. This allows any
partition to be scanned without scanning the entire disk and ensures
that the specified partition is free of bad blocks. As an example,
/dev/rra3h indicates the 
.B h 
partition of the third logical disk unit.
.PP
The following options may be set with 
.PN radisk:
.TP 15
.B \-c    
Clear a forced error indicator on the range of specified LBNs.  The
forced error condition indicates that the data in the disk block is
bad.  The disk block is good but the data can not be read without
getting a error detection code (EDC) error.  This option will cause the
forced error condition to be removed.  After the forced error indicator
is cleared, the EDC error will not be reported nor will the data be
marked as bad.  All indications that the data is corrupt are lost.  The
data should be restored either by manual methods or with the 
.PN restore
command. The 
.PN radisk 
command affects the integrity of the data on a disk and should be
followed by a file system restore if data is affected.
.TP 15
.B \-r
Replace a block on the disk.  The specified LBN will be replaced.
.TP 15
.B \-s
Start a scan for bad blocks on the specified area on a disk.  Bad blocks
are disk blocks that have data transfer errors to the extent that they
cannot be relied on.  When a bad block is found, it is replaced and the
bad block's LBN is reported.
.PP
.DT
.SH DIAGNOSTICS
The
.PN radisk
command generates messages when the user is not privileged, when the
LBN is not in the specified partition, and when the length exceeds the
size of the partition.
.SH RESTRICTIONS
The 
.B \-r 
option is supported only with those DSA disks which use host-initiated
dynamic bad block replacement. 
.SH SEE ALSO
dkio(4), ra(4), sdc(4), chpt(8), mount(8), restore(8)
.br
System Management Guide


etc/radisk.c
#ifndef lint
 static char *sccsid = "@(#)radisk.c	1.5    ULTRIX  1/9/87";
#endif

/************************************************************************
 *									*
 *			Copyright (c) 1986 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/
/***********************************************************************
 *
 *			Modification History
 *
 *
 **********************************************************************/

#define DEBUG


#include <stdio.h>
#include <sys/types.h>
#include <vax/mscp.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/param.h>
#include <ufs/fs.h>
#include <sys/dkio.h>
#include <sys/bbr.h>
#include <sys/ioctl.h>
#include <ctype.h>
#define BLOCKSIZE 512

char buf[BLOCKSIZE];

/*
 *	15-Dec-86	Robin
 *	Made changes to allow the scan option to work on controlers that don't
 *	have the mscp access command.  Also the special file must be the
 *	'c' partition except on the scan function.
 *
 */

/* This is a user level interface which allows a system administrator
 * to access the mscp disk and cause a block to be revectored, scan
 * the disk and report forced errors, clear forced errors, reset
 * the drive write protect software bit, and if I have time to print
 * out the controller type- micro code lev- ant the disk rct.
 *
 * The LBN used is the logical block number starting from the beginning
 * of the disk (512 byte blocks). If the input lbn is -1 then the beginning
 * of the specified partition is used.
 *
 * The length used is the number of blocks to process (512 byte blocks).
 * If the length is -1 then process to the end of the specified partition.
 *
 * The input flags are:
 *	-r	Causes the given lbn to be forced revectored by the controller.
 *		Only one lbn my be specified and the length is always 1.
 *	-s	Causes the controller to scan the disk for bad blocks.
 *	-c	Will clear forced error flags by reading the data ( the
 *		data is in an unknown state) and writing it back to the
 *		same block.  That will cause the data to be writen correctly
 *		or the block to be revectored and the data written in the
 *		new block.
 *	-u	This will reset the software write protect bit.
 *	-p	Will print the controller type, the micro code level and
 *		the rct (IF I HAVE TIME--A BELL)
 *
 * The special file input must be a raw device and the lbn input must be
 * in that partition.  If the length specified exceeds the partitions
 * end point then the length will be the partitions end.
 */



main(argc,argv)
int argc;
char *argv[];
{

struct pt partition;
struct dkacc data;
int ioctlstat;
int part_input;
int fd;
long part_size, part_start;
int flag;
/* static char usage[] = "usage: radisk -rcsup LBN [length] special_file"; */
static char usage[] = "usage: radisk -rcs LBN [length] special_file";
char *special;
char c_part[265];
long lbn;
int i;
int length = -1;
struct stat statbuf;
special = argv[argc - 1];
if(stat(special,&statbuf) == -1)
{
	fprintf(stderr,"Can't access special file %s\n",special);
	return(-1);
}
/* We have to open the c partition to get the partition table
 * information no matter what the input device partition was
*/
strcpy(c_part,special);
c_part[ strlen(c_part) - 1] = 'c';
argc--;
argv++;
if( argc < 3)
{
	fprintf(stderr,"%s\n",usage);
	return(-1);
}
if(*argv[0] != '-')
{
	fprintf(stderr,"%s\n",usage);
	return(-1);
}
switch (argv[0][1]) 
	{
			/* Toggle the sofwrae write protect bit in the drive */
	case 'u':
/*		data.dk_opcode = ACC_UNPROTECT; */
		return;
		break;
			/* Print out the controller, micro code lev. and rct */
	case 'p':
/*		data.dk_opcode = ACC_PRINT;*/
		return;
		break;
			/* Force clear the forced error bit */
	case 'c':
		data.dk_opcode = ACC_CLEAR;
		argc--;
		argv++;
		for (i=0; i < strlen(*argv); i++)
		{
			if (( isdigit(argv[0][i]) == 0 )&& (argv[0][i] != '-'))
			{
				fprintf(stderr,"lbn must be an integer\n");
				return(-1);
			}
		}
		lbn = atoi(*argv);
		argv++;
		for (i=0; i < strlen(*argv); i++)
		{
			if (( isdigit(argv[0][i]) == 0 )&& (argv[0][i] != '-'))
			{
				fprintf(stderr,"length must be an integer\n");
				return(-1);
			}
		}
		length = atoi(*argv);
		break;
			/* Force re-vector a disk block */
	case 'r':
		data.dk_opcode = ACC_REVEC;
		argc--;
		argv++;
		for (i=0; i < strlen(*argv); i++)
		{
			if (( isdigit(argv[0][i]) == 0 )&& (argv[0][i] != '-'))
			{
				fprintf(stderr,"lbn must be an integer\n");
				return(-1);
			}
		}
		lbn = atoi(*argv);
		break;
			/* Scan a disk for bad blocks, replace and report them */
	case 's':
		data.dk_opcode = ACC_SCAN;
		argc--;
		argv++;
		for (i=0; i < strlen(*argv); i++)
		{
			if (( isdigit(argv[0][i]) == 0 )&& (argv[0][i] != '-'))
			{
				fprintf(stderr,"lbn must be an integer\n");
				return(-1);
			}
		}
		lbn = atoi(*argv);
		argv++;
		for (i=0; i < strlen(*argv); i++)
		{
			if (( isdigit(argv[0][i]) == 0 )&& (argv[0][i] != '-'))
			{
				fprintf(stderr,"length must be an integer\n");
				return(-1);
			}
		}
		length = atoi(*argv);
	}
	if(data.dk_opcode != ACC_SCAN)
	if(special[strlen(special) - 1] != 'c')
	{
		fprintf(stderr,"special file for 'c' partition required\n");
		return;
	}

/* open the c partition to get the partition table */
if((fd=open(c_part,O_RDWR)) < 0)
{
	fprintf(stderr,"can't open %s\n",special);
	return(-1);
}
/* get the partition tabe info */
if(ioctl(fd, DIOCGETPT, (char *) &partition) < 0)
{
	perror("radisk: can't get partition table from driver");
	exit(-1);
}
close(fd);

/* open the partition that the user specified */
if((fd=open(special,O_RDWR)) < 0)
{
	fprintf(stderr,"can't open %s\n",special);
	return(-1);
}

/* Make sure the LBN is in the specified partition and that the length
 * does not exceed the end of that partition 
 */

part_input=(char) special[strlen(special) - 1];
part_size=partition.pt_part[(special[strlen(special) - 1] - 'a') ].pi_nblocks;
part_start=partition.pt_part[(special[strlen(special) - 1] - 'a') ].pi_blkoff;


if((lbn == -1) || (lbn == 0))
	lbn = part_start;

/* Make sure lbn is in the partition of the special device. */


if(lbn < part_start )
{
	printf("Input lbn %d is not in the %s partition\n",lbn,special);
	exit(-1);
}
if((part_start + part_size) < lbn)
{
	printf("Input lbn %d is not in %s the partition\n",lbn,special);
	exit(-1);
}
/*
calculate the offset from partition of special device input to the lbn,
if lbn is -1 offset = 0 and if length is -1 the go to end of partition.
*/
if ( length == -1)
	length = part_start + part_size - lbn;
if( (part_start + part_size) < (lbn + length) )
{
	length = part_start + part_size - lbn;
	printf("WARNING length input exceeded the partition, the length\n\twas changed to be the partitions end\n");
}

/* load data for iooctl call and then process the ioctl */

data.dk_lbn = lbn;
data.dk_length = length  * 512 ;
data.dk_status = -1;

switch ( data.dk_opcode)
{
	case ACC_CLEAR:
		ioctlstat=do_clear(fd,lbn - part_start,length);
		break;
	case ACC_SCAN:
		fprintf(stderr,"Scan started :\n\n");
		ioctlstat=do_scan(fd,&data);
		fprintf(stderr,"\n");
		break;
	case ACC_REVEC:
		ioctlstat=ioctl(fd, DKIOCACC, (char *) &data);
		break;
/*	
	case ACC_PRINT:
		ioctlstat=ioctl(fd, DKIOCACC, (char *) &data);
		break;
	case ACC_UNPROTECT:
		ioctlstat=ioctl(fd, DKIOCACC, (char *) &data);
		break;
*/
}
	if (ioctlstat == -1) perror("radisk");
}

/*
 * do_clear reads a block of data and writes it back out to the same place on
 * the disk.  The write will force the data to a good block if the block
 * is a bad block or it will clear a forced error flag if it was set.  If
 * the forced error was set the data is in an unknkown state.
 *
 * This is not a good thing to do unless you REALY know what you are up to
 * the file system should be restored after using this because the data
 * will be in an unknown state.
 */

do_clear(fd,lbn,length)
int fd;
long lbn, length;
{
	int charcnt;
	for (;length > 0;length--)
	{
		if((lseek(fd,(lbn*512),L_SET)) == -1)
		{
			perror("radisk: seek error, process aborted \n");
			exit(-1);
		}
		charcnt=read(fd, buf, BLOCKSIZE);
		if((lseek(fd,(lbn * 512),L_SET)) == -1)
		{
			perror("radisk: seek error, process aborted \n");
			return(-1);
		}
		if((charcnt=write(fd, buf, BLOCKSIZE)) == -1)
		{
			perror("radisk: write error, process aborted \n");
			return(-1);
		}
		lbn++;
	}
return(0);
}

/* Send the mscp_scan command to the controller.  If it finds a bad block
 * it will return a bad block reported after it has revectored the block.
 * We should restart the scan, this will continue until the disk is compleatly
 * checked.  The scan starts up where the use indicated and not where
 * the error was found becaus we want to end with an error free scan of
 * the entire disk section. I think errors should be far and few between
 * so the compleat restart will not cause large delays.  If thats not
 * true we may want to change the logic to restart near the found bad block.
 */

do_scan(fd,data)
int fd;
struct dkacc *data;
{
	int ioctlstat;
	struct dkacc localdata;
	localdata.dk_opcode = data->dk_opcode;
	localdata.dk_lbn = data->dk_lbn;
	localdata.dk_flags = data->dk_flags;
	localdata.dk_status = data->dk_status;
	localdata.dk_length = data->dk_length;

	ioctlstat=ioctl(fd, DKIOCACC, &localdata);
	if(ioctlstat == -1) 
		return(ioctlstat);
	if(localdata.dk_status == NO_ACCESS)
	{
		do_read(fd,data->dk_lbn,data->dk_length);
		return;
	}
	if(localdata.dk_status != M_ST_SUCC)
	{
		/* If a bad block is found, print it's LBN and restart the scan. */
		if(localdata.dk_flags & M_EF_BBLKR)
		{
			fprintf(stderr,"%d replaced\n",localdata.dk_lbn);
			do_scan(fd,data);
		}
		/* On forced error report that it was found, skip it and keep going*/
		if(localdata.dk_status & M_DT_FEM)
		{
		/* compute the lbn that had the error from the byte count returned */
			localdata.dk_lbn = data->dk_lbn + (long) (localdata.dk_length / BLOCKSIZE);
			fprintf(stderr,"forced error set at LBN %d \n",localdata.dk_lbn);
			localdata.dk_opcode = data->dk_opcode;
			localdata.dk_lbn++;
			localdata.dk_flags = data->dk_flags;
			localdata.dk_status = data->dk_status;
			localdata.dk_length = data->dk_length - (localdata.dk_length) - BLOCKSIZE;
			do_scan(fd,&localdata);
		}
		else
		{
			fprintf(stderr,"Aborted mscp status is o%o flags are o%o\n",
				localdata.dk_status,
				localdata.dk_flags);
		}
	}
	return(ioctlstat);
}

/* do_read is a function that simulates the scan option for disks that
   do not support the mscp access function.  Just read everything and let the
   driver report any problems.
*/

do_read(fd,lbn,length)
int fd;
long lbn, length;
{
	char buffer[BLOCKSIZE];
	int charcnt;
	/* change length in bytes to length in sectors */
	length = (length/512)+1;
	for (;length > 0;length--)
	{
		if((lseek(fd,(lbn*512),L_SET)) == -1)
		{
			perror("radisk: seek error, process aborted \n");
			exit(-1);
		}
		read(fd, buffer, BLOCKSIZE);
		lbn++;
	}
}
